### PR TITLE
Feature: node storage optimization

### PR DESCRIPTION
**Pull request by liss-h**
This PR attempts to optimize node-storage size and access pattern.

Size is reduced by:
  - using a custom string type that does not store the capacity (minimal improvement, but storing the capacity is not necessary)
  - using a `std::vector` instead of `sparse_map` for the forward lookup direction (not sure how much this does)

Access pattern is improved by:
  - increased data locality achieved by flattening the forward access structure (lookup by id now takes 3 indirections instead of 5)

### CODE REVIEW

**Comment by liss-h** [comment]
TODO:
  - [x] fix the serialization test case (it has the triples in the wrong order now because different order in storage)

File: .github/workflows/run_test_and_examples.yml [hunk 1]
```diff
@@ -8,8 +8,6 @@ jobs:
           - os: ubuntu-22.04
             compiler: gcc-13
 
-          - os: ubuntu-22.04
-            compiler: clang-15
```
Comment (bigerl) [inline .github/workflows/run_test_and_examples.yml:12]: why was clang-15 removed?

File: src/rdf4cpp/rdf/storage/node/identifier/NodeID.hpp [hunk 2]
```diff
@@ -76,6 +76,9 @@ class __attribute__((__packed__)) NodeID {
      */
     [[nodiscard]] constexpr LiteralType literal_type() const noexcept { return LiteralType::from_underlying(literal_.literal_type); }
     [[nodiscard]] constexpr uint64_t value() const noexcept { return value_; }
+
+    explicit operator uint64_t() const noexcept { return value_; }
+
     //    void value(uint64_t val) { assert(val < (1UL << 48)); value_ = val; }
```
Comment (bigerl) [inline src/rdf4cpp/rdf/storage/node/identifier/NodeID.hpp]: remove the comment

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        size_t hash; // hash of the Value being looked up
+        id_type id;  // id into forward_
+    };
+
+    struct backward_key_equal {
+        using is_transparent = void;
+
+        using forward_const_pointer = typename std::allocator_traits<allocator_type>::template rebind_traits<forward_type>::const_pointer;
+
+        forward_const_pointer forward_ptr; // pointer to forward_ to be able to check for equality between View and Value
```
Comment (bigerl) [inline src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp]:
```suggestion
        forward_const_pointer forward_ptr; ///< pointer to forward_ to be able to check for equality between View and Value
```

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    backward_type backward_; // backward_[view] stores Id into forward_
+    [[no_unique_address]] allocator_type alloc_;
+
+    [[nodiscard]] static constexpr size_type to_index(id_type id) noexcept {
+        return static_cast<size_type>(id) - 1;
+    }
+
+    [[nodiscard]] static constexpr id_type to_id(size_type ix) noexcept {
+        return static_cast<id_type>(ix + 1);
+    }
```
Comment (bigerl) [inline src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp]: Please add a docstring to these

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    BiDirFlatMap &operator=(BiDirFlatMap const &) = delete;
+    BiDirFlatMap &operator=(BiDirFlatMap &&) = delete;
+
+    [[nodiscard]] size_type size() const noexcept {
+        return forward_.size();
+    }
+
+    void shrink_to_fit() {
+        forward_.shrink_to_fit();
+        // TODO backward_?
```
Comment (bigerl) [inline src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp]: rehash(0)

Comment (bigerl) [inline src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp]:
```suggestion
        backward_.rehash(0);
```

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+        auto it = backward_.find(view);
+        if (it == backward_.end()) {
+            return Id{};
+        }
+
+        return it->id;
+    }
+
+    void reserve(id_type const min_id) {
+        forward_.resize(to_index(min_id));
```
Comment (bigerl) [inline src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp]: this might delete entries. only `min_id >= forward.size()` should be valid.

Comment (bigerl) [inline src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp]: also do `bachward_.reserve(min_id);`

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+    [[nodiscard]] static constexpr size_type to_index(id_type id) noexcept {
+        return static_cast<size_type>(id) - 1;
+    }
+
+    [[nodiscard]] static constexpr id_type to_id(size_type ix) noexcept {
+        return static_cast<id_type>(ix + 1);
+    }
+
+public:
+    explicit BiDirFlatMap(hasher const &hash = hasher{},
```
Comment (bigerl) [inline src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp:111]: Please add docstrings to the public methods and ctors

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/ConstString.hpp [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        // prevent copy constructors from being accidentally called
+        BasicConstString(BasicConstString const &) = delete;
+        BasicConstString &operator=(BasicConstString const &) = delete;
+
+        BasicConstString(BasicConstString &&other) noexcept : data_{std::exchange(other.data_, nullptr)},
+                                                              size_{std::exchange(other.size_, 0)},
+                                                              alloc_{std::move(other.alloc_)} {
+        }
+
+        BasicConstString &operator=(BasicConstString &&other) noexcept(std::allocator_traits<Allocator>::propagate_on_container_move_assignment::value
```
Comment (bigerl) [inline src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/ConstString.hpp:49]: nicely done

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/ConstString.hpp [hunk 9]
```diff
@@ -0,0 +1,103 @@
+#ifndef RDF4CPP_REFERENCENODESTORAGE_CONSTSTRING_HPP
+#define RDF4CPP_REFERENCENODESTORAGE_CONSTSTRING_HPP
+
+#include <string>
+
+namespace rdf4cpp::rdf::storage::node::reference_node_storage::detail {
+
+    template<typename Char, typename Traits = std::char_traits<Char>, typename Allocator = std::allocator<Char>>
+    struct BasicConstString {
```
Comment (bigerl) [inline src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/ConstString.hpp:13]: please add at least a docstring to the class

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/SpecializationDetail.hpp [hunk 10]
```diff
@@ -20,23 +20,42 @@ constexpr Acc tuple_fold_impl(std::index_sequence<Ixs...>, Tuple const &tuple, A
...
+    return tuple_fold_impl(std::make_index_sequence<std::tuple_size_v<std::remove_cvref_t<Tuple>>>{}, tuple, std::forward<Acc>(init), std::forward<FoldF>(f));
+}
+
+template<typename Tuple, typename F>
+constexpr void tuple_for_each(Tuple &&tuple, F &&f) {
+    return tuple_for_each_impl(std::make_index_sequence<std::tuple_size_v<std::remove_cvref_t<Tuple>>>{}, std::forward<Tuple>(tuple), std::forward<F>(f));
+}
+
+template<typename Tuple, typename F>
+constexpr void tuple_type_for_each(F &&f) {
```
Comment (bigerl) [inline src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/SpecializationDetail.hpp:49]: we could think about moving the tuple functions to dice template library

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/SpecializationDetail.hpp [hunk 11]
```diff
@@ -20,23 +20,42 @@ constexpr Acc tuple_fold_impl(std::index_sequence<Ixs...>, Tuple const &tuple, A
...
+    return tuple_type_for_each_impl<Tuple>(std::make_index_sequence<std::tuple_size_v<Tuple>>{}, std::forward<F>(f));
 }
 
 template<typename Tuple>
 static consteval std::array<bool, 1 << identifier::LiteralType::width> make_storage_specialization_lut() noexcept {
```
Comment (bigerl) [inline src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/SpecializationDetail.hpp:63]: Please add a docstring.

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/UnsyncNodeTypeStorage.hpp [hunk 12]
```diff
@@ -22,15 +23,12 @@ struct UnsyncNodeTypeStorage {
     struct DefaultBackendTypeEqual {
         using is_transparent = void;
 
-        bool operator()(Backend const *lhs, Backend const *rhs) const noexcept {
```
Comment (bigerl) [inline src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/UnsyncNodeTypeStorage.hpp:25]: did you remove this because it is no longer required?

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/UnsyncNodeTypeStorage.hpp [hunk 13]
```diff
@@ -68,15 +63,25 @@ struct UnsyncNodeTypeStorage {
...
-    };
+    }
+
+    static BackendId to_backend_id(identifier::NodeID const id) noexcept {
+        if constexpr (requires { Backend::to_backend_id(id); }) {
+            return Backend::to_backend_id(id);
+        } else {
+            return id;
+        }
+    }
```
Comment (bigerl) [inline src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/UnsyncNodeTypeStorage.hpp]: add docstrings

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+    }
+
+    void reserve(id_type const min_id) {
+        forward_.resize(to_index(min_id));
+    }
+
+    [[nodiscard]] id_type insert_assume_not_present(view_type const &view) {
+        assert(lookup_id(view) == Id{});
+
+        auto const assigned_id = to_id(forward_.size());
```
Comment (bigerl) [inline src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp]: This is a general design issue: if we create n + 1 entries, remove the first n entries and add another `n` entries we will end up with 2n + 1 slots used. 

This is not easy to solve. The clean solution would be managing holes (as in files). The easy solution are tombstones. I would suggest a vector of `size_t`s where the n-th bit tells if the n-th slot is free.

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/SyncReferenceNodeStorageBackend.cpp [hunk 15]
```diff
@@ -4,25 +4,26 @@
 namespace rdf4cpp::rdf::storage::node::reference_node_storage {
 
 SyncReferenceNodeStorageBackend::SyncReferenceNodeStorageBackend() noexcept {
```
Comment (bigerl) [inline src/rdf4cpp/rdf/storage/node/reference_node_storage/SyncReferenceNodeStorageBackend.cpp:6]: side note: how about a common parent class for Sync/Unsync Ref Node Storage?

File: tests/serializer/tests_Serialization.cpp [hunk 16]
```diff
@@ -18,35 +18,43 @@ TEST_CASE("Literal short type") {
     CHECK(ser.finalize()=="\"2042-05-04\"^^xsd:date\ntrue\n\"4.0E0\"^^xsd:float\n\"4\"^^xsd:unsignedByte\n4");
 }
 
-template<writer::OutputFormat F>
-bool serialize(const Quad& q, void *buffer, writer::Cursor &cursor, writer::FlushFunc flush, writer::SerializationState* state) {
```
Comment (bigerl) [inline tests/serializer/tests_Serialization.cpp]: these changes seem totally unrelated

**Review by bigerl — CHANGES_REQUESTED** [review]
thx for the PR. I like the proposed method because it makes things much cleaner. 

There is one major issue though with the design. See my comment in BiDirFlatMap.

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp (hunk 14 above)
Comment (liss-h) [reply src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp]: We had almost the same issue before (i.e. ids not being reused) just that it didn't occupy additional space. Didn't really think about that.

So I see two options:
1. keep the increasing id behaviour as is, apply an offset to the vector index so that we do not need to store the empty range at the start
2. Solve both problems by reusing ids

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/SyncReferenceNodeStorageBackend.cpp (hunk 15 above)
Comment (liss-h) [reply src/rdf4cpp/rdf/storage/node/reference_node_storage/SyncReferenceNodeStorageBackend.cpp:6]: I tried that in the original PR that introduced `UnsyncReferenceNodeStorageBackend`, but I stopped at some point because I decided that the additional effort was not worth it

File: tests/serializer/tests_Serialization.cpp (hunk 16 above)
Comment (liss-h) [reply tests/serializer/tests_Serialization.cpp]: 😅 I guess that happens if you work on two PRs at once. Those changes were meant for the other one

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/SpecializationDetail.hpp (hunk 10 above)
Comment (liss-h) [reply src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/SpecializationDetail.hpp:49]: Yeah, probably makes sense

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp (hunk 5 above)
Comment (liss-h) [reply src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp]: Is the zero just a placeholder value that tells it to clean up?

File: .github/workflows/run_test_and_examples.yml (hunk 1 above)
Comment (liss-h) [reply .github/workflows/run_test_and_examples.yml:12]: Supporting it was additional effort because it apparently doesn't generate automatic constructors properly. See https://github.com/rdf4cpp/rdf4cpp/actions/runs/8063550056/job/22025596346

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp (hunk 5 above)
Comment (bigerl) [reply src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp]: yes, see https://en.cppreference.com/w/cpp/container/unordered_map/rehash

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp (hunk 14 above)
Comment (bigerl) [reply src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp]: I would reuse IDs. Let's have a short call about how to solve this tomorrow.

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/SpecializationDetail.hpp (hunk 10 above)
Comment (liss-h) [reply src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/SpecializationDetail.hpp:49]: I will create an issue in template-lib and do it at some point later

File: tests/serializer/tests_Serialization.cpp (hunk 16 above)
Comment (liss-h) [reply tests/serializer/tests_Serialization.cpp]: I will just keep it in here, my intention is to merge the other PR first anyway

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp (hunk 6 above)
Comment (liss-h) [reply src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp]: added check

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/SpecializationDetail.hpp (hunk 10 above)
Comment (liss-h) [reply src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/SpecializationDetail.hpp:49]: https://github.com/dice-group/dice-template-library/issues/33

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp (hunk 14 above)
Comment (liss-h) [reply src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp]: Implemented

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp (hunk 7 above)
Comment (liss-h) [reply src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp:111]: Left out ctors because there isn't really anything to document

**Comment by liss-h** [comment]
Coverage is broken somehow, no idea what thats about

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/IndexFreeList.hpp [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+        auto const last_occupied = std::find_if(occupied_bitmap_.rbegin(), occupied_bitmap_.rend(), [](auto const bitmap) noexcept {
+            return bitmap != 0;
+        });
+
+        auto const unused_size = std::distance(occupied_bitmap_.rbegin(), last_occupied);
+        occupied_bitmap_.resize(occupied_bitmap_.size() - unused_size);
+    }
+
+    /**
+     * Marks all slots up-to (not-including) the given index as occupied
```
Comment (bigerl) [inline src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/IndexFreeList.hpp:104]: the implementation does not cover the case ix = 70 (>63) but there is a true at 130 (>127).

what is missing is a fill for slots before slot_ix if the condition in line 106 is not met. 

I think this is not practically relevant for our use-case but also very unexpected behavior. So, I would feel safer if it would be fixed.

**Review by bigerl — CHANGES_REQUESTED** [review]
Changes look good. There is just one little corner case I found.

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/IndexFreeList.hpp (hunk 17 above)
Comment (liss-h) [reply src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/IndexFreeList.hpp:104]: Fixed this and another bug related to `occupy_until` (that also doesn't occur with our usecase, but can't hurt to fix it)

File: src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/IndexFreeList.hpp [hunk 18]
```diff
@@ -102,14 +104,19 @@ struct IndexFreeList {
...
+            occupied_bitmap_.resize(slot_ix + 1, all_ones);
+            occupied_bitmap_[slot_ix] = (1ul << offset) - 1;
+        } else {
+            std::fill_n(occupied_bitmap_.begin(), slot_ix, all_ones);
+            occupied_bitmap_[slot_ix] |= (1ul << offset) - 1;
         }
 
-        occupied_bitmap_[slot_ix] = (1ul << offset) - 1;
-        next_free_ix_ = ix;
+        next_free_ix_ = search_next_free(slot_ix, offset);
```
Comment (bigerl) [inline src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/IndexFreeList.hpp:119]: good catch

**Review by bigerl — APPROVED** [review]
